File: src/benchmarks/uncapped-fps.ts
import {
  BLACK,
  BLUE,
  DARKBLUE,
  DARKGRAY,
  GREEN,
  ORANGE,
  RAYWHITE,
  RED,
  beginDrawing,
  clearBackground,
  closeWindow,
  drawCircle,
  drawFPS,
  drawRectangle,
  drawText,
  endDrawing,
  getTime,
  initWindow,
  setTargetFPS,
  windowShouldClose,
} from "../raylib";

const screenWidth = 1280;
const screenHeight = 720;
const warmupSeconds = 2;
const baselineSeconds = 5;
const stressSeconds = 5;
const stressRectangleCount = 1000;

interface BenchmarkResult {
  frames: number;
  seconds: number;
  minimumFrameSeconds: number;
  maximumFrameSeconds: number;
}

function drawBaselineScene(frame: number): void {
  const rectangleX = 80 + (frame % 900);
  const circleX = 1180 - (frame % 1100);

  beginDrawing();
  clearBackground(RAYWHITE);
  drawText("Uncapped baseline: starter-sized scene", 20, 20, 24, BLACK);
  drawText("No target FPS and no VSYNC hint", 20, 55, 18, DARKGRAY);
  drawText("This phase draws three strings, one rectangle, and one circle.", 20, 82, 18, DARKGRAY);
  drawRectangle(rectangleX, 330, 110, 55, BLUE);
  drawCircle(circleX, 480, 24, RED);
  drawFPS(20, 120);
  endDrawing();
}

function drawStressScene(frame: number): void {
  const offset = frame % 24;

  beginDrawing();
  clearBackground(RAYWHITE);

  let index = 0;
  while (index < stressRectangleCount) {
    const column = index % 50;
    const row = Math.floor(index / 50);
    const x = column * 26 + offset - 24;
    const y = row * 31 + 96;
    const colorIndex = index % 4;
    if (colorIndex === 0) drawRectangle(x, y, 20, 20, BLUE);
    else if (colorIndex === 1) drawRectangle(x, y, 20, 20, GREEN);
    else if (colorIndex === 2) drawRectangle(x, y, 20, 20, ORANGE);
    else drawRectangle(x, y, 20, 20, RED);
    index += 1;
  }

  drawRectangle(0, 0, screenWidth, 90, RAYWHITE);
  drawText("Uncapped draw-call stress: 1,000 rectangles per frame", 20, 18, 24, DARKBLUE);
  drawFPS(20, 52);
  endDrawing();
}

function runWarmup(): boolean {
  const startedAt = getTime();
  let frame = 0;
  while (!windowShouldClose() && getTime() - startedAt < warmupSeconds) {
    drawBaselineScene(frame);
    frame += 1;
  }
  return !windowShouldClose();
}

function runBenchmark(duration: number, stress: boolean): BenchmarkResult {
  const startedAt = getTime();
  let frame = 0;
  let minimumFrameSeconds = 1000000000;
  let maximumFrameSeconds = 0;

  while (!windowShouldClose()) {
    const frameStartedAt = getTime();
    if (stress) drawStressScene(frame);
    else drawBaselineScene(frame);
    const frameSeconds = getTime() - frameStartedAt;

    if (frameSeconds < minimumFrameSeconds) minimumFrameSeconds = frameSeconds;
    if (frameSeconds > maximumFrameSeconds) maximumFrameSeconds = frameSeconds;
    frame += 1;

    if (getTime() - startedAt >= duration) break;
  }

  return {
    frames: frame,
    seconds: getTime() - startedAt,
    minimumFrameSeconds,
    maximumFrameSeconds,
  };
}

function printResult(name: string, result: BenchmarkResult): void {
  if (result.frames === 0) {
    console.log(`${name}: no frames recorded`);
    return;
  }

  const averageFPS = result.frames / result.seconds;
  const averageFrameMilliseconds = (result.seconds * 1000) / result.frames;
  console.log(`${name}:`);
  console.log(`  frames: ${result.frames} in ${result.seconds.toFixed(3)} seconds`);
  console.log(`  average: ${averageFPS.toFixed(1)} FPS (${averageFrameMilliseconds.toFixed(3)} ms/frame)`);
  console.log(`  best frame: ${(result.minimumFrameSeconds * 1000).toFixed(3)} ms`);
  console.log(`  worst frame: ${(result.maximumFrameSeconds * 1000).toFixed(3)} ms`);
}

console.log("scriptc-raylib uncapped FPS benchmark");
console.log(`Resolution: ${screenWidth}x${screenHeight}`);
console.log(`Warm-up: ${warmupSeconds}s; samples: ${baselineSeconds}s baseline + ${stressSeconds}s stress`);

initWindow(screenWidth, screenHeight, "scriptc-raylib uncapped FPS benchmark");
setTargetFPS(0);

if (runWarmup()) {
  const baseline = runBenchmark(baselineSeconds, false);
  if (!windowShouldClose()) {
    const stress = runBenchmark(stressSeconds, true);
    closeWindow();
    printResult("Baseline scene", baseline);
    printResult(`Stress scene (${stressRectangleCount} rectangles/frame)`, stress);
  } else {
    closeWindow();
    console.log("Benchmark stopped before the stress phase.");
    printResult("Baseline scene", baseline);
  }
} else {
  closeWindow();
  console.log("Benchmark stopped during warm-up.");
}

File: benchmarks/uncapped-fps.c
#include "raylib.h"

#include <stdbool.h>
#include <stdio.h>

enum {
    SCREEN_WIDTH = 1280,
    SCREEN_HEIGHT = 720,
    STRESS_RECTANGLE_COUNT = 1000,
};

static const double WARMUP_SECONDS = 2.0;
static const double BASELINE_SECONDS = 5.0;
static const double STRESS_SECONDS = 5.0;

typedef struct BenchmarkResult {
    int frames;
    double seconds;
    double minimum_frame_seconds;
    double maximum_frame_seconds;
} BenchmarkResult;

static void draw_baseline_scene(int frame) {
    const int rectangle_x = 80 + (frame % 900);
    const int circle_x = 1180 - (frame % 1100);

    BeginDrawing();
    ClearBackground(RAYWHITE);
    DrawText("Uncapped baseline: starter-sized scene", 20, 20, 24, BLACK);
    DrawText("No target FPS and no VSYNC hint", 20, 55, 18, DARKGRAY);
    DrawText("This phase draws three strings, one rectangle, and one circle.", 20, 82, 18, DARKGRAY);
    DrawRectangle(rectangle_x, 330, 110, 55, BLUE);
    DrawCircle(circle_x, 480, 24.0f, RED);
    DrawFPS(20, 120);
    EndDrawing();
}

static void draw_stress_scene(int frame) {
    const int offset = frame % 24;

    BeginDrawing();
    ClearBackground(RAYWHITE);

    int index = 0;
    while (index < STRESS_RECTANGLE_COUNT) {
        const int column = index % 50;
        const int row = index / 50;
        const int x = column * 26 + offset - 24;
        const int y = row * 31 + 96;
        const int color_index = index % 4;
        if (color_index == 0) DrawRectangle(x, y, 20, 20, BLUE);
        else if (color_index == 1) DrawRectangle(x, y, 20, 20, GREEN);
        else if (color_index == 2) DrawRectangle(x, y, 20, 20, ORANGE);
        else DrawRectangle(x, y, 20, 20, RED);
        index += 1;
    }

    DrawRectangle(0, 0, SCREEN_WIDTH, 90, RAYWHITE);
    DrawText("Uncapped draw-call stress: 1,000 rectangles per frame", 20, 18, 24, DARKBLUE);
    DrawFPS(20, 52);
    EndDrawing();
}

static bool run_warmup(void) {
    const double started_at = GetTime();
    int frame = 0;
    while (!WindowShouldClose() && GetTime() - started_at < WARMUP_SECONDS) {
        draw_baseline_scene(frame);
        frame += 1;
    }
    return !WindowShouldClose();
}

static BenchmarkResult run_benchmark(double duration, bool stress) {
    const double started_at = GetTime();
    int frame = 0;
    double minimum_frame_seconds = 1000000000.0;
    double maximum_frame_seconds = 0.0;

    while (!WindowShouldClose()) {
        const double frame_started_at = GetTime();
        if (stress) draw_stress_scene(frame);
        else draw_baseline_scene(frame);
        const double frame_seconds = GetTime() - frame_started_at;

        if (frame_seconds < minimum_frame_seconds) minimum_frame_seconds = frame_seconds;
        if (frame_seconds > maximum_frame_seconds) maximum_frame_seconds = frame_seconds;
        frame += 1;

        if (GetTime() - started_at >= duration) break;
    }

    const BenchmarkResult result = {
        .frames = frame,
        .seconds = GetTime() - started_at,
        .minimum_frame_seconds = minimum_frame_seconds,
        .maximum_frame_seconds = maximum_frame_seconds,
    };
    return result;
}

static void print_result(const char *name, BenchmarkResult result) {
    if (result.frames == 0) {
        printf("%s: no frames recorded\n", name);
        return;
    }

    const double average_fps = (double)result.frames / result.seconds;
    const double average_frame_milliseconds = (result.seconds * 1000.0) / (double)result.frames;
    printf("%s:\n", name);
    printf("  frames: %d in %.3f seconds\n", result.frames, result.seconds);
    printf("  average: %.1f FPS (%.3f ms/frame)\n", average_fps, average_frame_milliseconds);
    printf("  best frame: %.3f ms\n", result.minimum_frame_seconds * 1000.0);
    printf("  worst frame: %.3f ms\n", result.maximum_frame_seconds * 1000.0);
}

int main(void) {
    printf("native C raylib uncapped FPS benchmark\n");
    printf("Resolution: %dx%d\n", SCREEN_WIDTH, SCREEN_HEIGHT);
    printf(
        "Warm-up: %.0fs; samples: %.0fs baseline + %.0fs stress\n",
        WARMUP_SECONDS,
        BASELINE_SECONDS,
        STRESS_SECONDS
    );

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, "native C raylib uncapped FPS benchmark");
    SetTargetFPS(0);

    if (run_warmup()) {
        const BenchmarkResult baseline = run_benchmark(BASELINE_SECONDS, false);
        if (!WindowShouldClose()) {
            const BenchmarkResult stress = run_benchmark(STRESS_SECONDS, true);
            CloseWindow();
            print_result("Baseline scene", baseline);
            print_result("Stress scene (1000 rectangles/frame)", stress);
        } else {
            CloseWindow();
            printf("Benchmark stopped before the stress phase.\n");
            print_result("Baseline scene", baseline);
        }
    } else {
        CloseWindow();
        printf("Benchmark stopped during warm-up.\n");
    }

    return 0;
}
